### DISCUSSION TITLE
Define supported platforms for turborepo

### DISCUSSION THREAD

**Discussion by gsoltis** [vercel/turborepo#1891] [RFC]
We are intending on enabling `CGO` across the board for `turbo` so that we can start interfacing with native libraries. This has the consequence of needing to put more effort into cross-compiling our Go code. 

To facilitate cross compilation, we would like to clearly specify which platforms `turbo` supports and intends to support.

The current proposal is to support each of:
 - linux
 - macos
 - windows

on
 - amd64
 - arm64

Anyone using any of the above combinations will be unaffected by this change.

If there are `turbo` users not using a platform listed above, we'd like to hear from you. `turbo` is currently built for a variety of more esoteric platforms as well, but we have not identified usage for them via our remote cache stats. However, it is possible that there are users not using remote caching and building on a platform that is not covered by the above, and we'd like to make sure they have a path forward if at all possible.

Packages we would no longer build:
- [turbo-android-arm64](https://www.npmjs.com/package/turbo-android-arm64)
- [turbo-freebsd-64](https://www.npmjs.com/package/turbo-freebsd-64)
- [turbo-freebsd-arm64](https://www.npmjs.com/package/turbo-freebsd-arm64)
- [turbo-linux-32](https://www.npmjs.com/package/turbo-linux-32)
- [turbo-linux-arm](https://www.npmjs.com/package/turbo-linux-arm) (this is 32 bit, we would still build turbo-linux-arm64)
- [turbo-linux-mips64le](https://www.npmjs.com/package/turbo-linux-mips64le)
- [turbo-linux-ppc64le](https://www.npmjs.com/package/turbo-linux-ppc64le)
- [turbo-windows-32](https://www.npmjs.com/package/turbo-windows-32)

**Comment by TheKnarf** [comment]
Linux ARM 32bit would be useful for building and deploying apps that run on a Raspberry PI. While it is not the most powerful platform it is an easy way to have a small cluster at home for testing purposes. Raspberry PI have gotten experimental support for 64bit, but most people still run 32bit since it have better stability.

**Reply by nathanhammond** [reply]
Are you presently using Turborepo on a raspberry pi for builds? My general estimation is that it is primarily used as a deploy target. In other words, the build that uses Turborepo would not be run on it. Instead, the output would be sent to it.

**Reply by TheKnarf** [reply]
I'm both building stuff on and running stuff on Raspberry PI's from time to time. To deploy stuff to a Raspberry PI in a Docker container I also use `docker buildx`, I belive `docker buildx` may emulate linux arm 32bit to create a Docker container for that platform, but I'm not entirely sure.

**Reply by nathanhammond** [reply]
I believe that workflow should continue to work with Docker using the host env to run the build, but do let me know if it stops working for some reason!

**Reply by gsoltis** [reply]
@TheKnarf can you try installing the platform-specific `turbo` package: https://www.npmjs.com/package/turbo-linux-arm64

If you can run the binary from that package on the raspberry pi, you should be good to go. Even just running `turbo bin` should do it to confirm.

**Comment by darkbasic** [comment]
Any chance to revert this? We need both linux-arm and linux-ppc64le and you caught us off guard by not respecting semantic versioning with this change: https://github.com/zwave-js/node-zwave-js/issues/5164
linux-arm is because of the Raspberry Pi and ppc64le is my main workstation for fullstack web development. Ever heard of Raptor Computing Talos 2? It's the only fully-free-down-to-the-firmware system in the world and it's based on power 9.

**Reply by nathanhammond** [reply]
Heya @darkbasic! We're sorry for messing up your workflow. The two platforms you report needing are _likely_ not a super-heavy lift. We've migrated to CGO as an enabler for an incremental Rust migration which means that we are no longer able to leverage the built-in Go cross-compilation strategy we were using previously.

We're willing to support both of those platforms, but it's low priority for us to figure out the magic incantation for our build container that would enable targeting those architectures: https://github.com/vercel/turborepo/blob/main/buildcontainer/Dockerfile

Would you be willing to help us figure out the magic build toolchain to support both of those architectures?

**Reply by darkbasic** [reply]
What's the problem with your current toolchain? I see there is already an armv7 toolchain [here](https://github.com/mstorsjo/llvm-mingw/releases) so it must be broken somehow, on the contrary I see no ppc64le toolchain. Maybe @madscientist159 can point us to one readily available.

**Reply by madscientist159** [reply]
> What's the problem with your current toolchain? I see there is already an armv7 toolchain [here](https://github.com/mstorsjo/llvm-mingw/releases) so it must be broken somehow, on the contrary I see no ppc64le toolchain. Maybe @madscientist159 can point us to one readily available.

I think that toolchain is only being brought in for Windows support on ARM.  We don't need Windows support for ppc64el.

ppc64el is very well supported on Debian, I suspect all you need to do is copy all the `arm64` lines, substituting `ppc64el`.  Then, alter the mingw section to not download the mingw toolchain when the architecture is `ppc64el`.  From what I can see that should at least get us in the ballpark?

**Comment by anthonyshew** [comment]
https://turbo.build/repo/docs/support
